FEAT: Explore drawer panel

me-imfhd opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This adds an explore-drawer component inspired by track-challenge-drawer.
## Description
<!--- Describe your changes in detail -->
1) A separate NavLink component for Explore is implemented. When on a path that starts with /challenge, it functions as a sheet drawer, similar to the track-challenge one.
2) The panel defaults to displaying popular challenges. Subsequently, challenges can be filtered based on difficulty.
3) A progress bar is added for each difficulty, resembling the profiles one.

To get a better understanding of the visual appearance, refer to the  video.

To wrap it up, the code is working, but it might not be complete. Please take a look and let me know if there's anything I can do to make it better.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->https://github.com/typehero/typehero/issues/1335
Closes #

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->

## Screenshots/Video (if applicable):
[Screencast from 27-12-23 06:34:14 PM IST.webm](https://github.com/typehero/typehero/assets/114667178/216ae178-9fdc-44b3-ad70-a6a236dbc3e5)

jhechtf commented:
What does this look like in a full window? It appears that it takes up the whole screen but I'm sure that's likely just due to cropping?

me-imfhd commented:
Added three contexts: 
1) useChallengeRouteData, to get current challenge on the route throughtout the app,
2) & 3) two new context for setting track and for setting sortKeys, so we don't lose state and avoid prop drilling and clean code

added previous and next buttons for challenges track

user can set there track automatically by selecting the popular, easy, medium... and so on, can be altered to use tags too

to understand better please watch the video attached.
[Screencast from 29-12-23 07:59:33 PM IST.webm](https://github.com/typehero/typehero/assets/114667178/dc6828b1-4978-4617-8f25-0cc86e5c9283)

me-imfhd commented:
> What does this look like in a full window? It appears that it takes up the whole screen but I'm sure that's likely just due to cropping?

yes sir you are right!

bautistaaa commented:
on my todos to look at this soon. been out cause family is sick

jhechtf commented:
I pushed your branch to a preview deployment, but I'm having a bit of trouble getting the drawer to show up. Do you mind writing down the steps to get it to show up?

jhechtf commented:
I'm seeing this weird glitch when I attempt to change the dropdown for the challenges

https://github.com/typehero/typehero/assets/8152471/343ca673-5483-451b-bdba-117a266cfb5e

jhechtf commented:
I would probably not have that dropdown change the page the user is on. The user should be navigated somewhere _if_ they click on one of those challenges. I hadn't noticed I was being moved around the site when changing that dropdown.

me-imfhd commented:
@jhechtf i made the change to not push to the first challenge when selecting another track.
Can you review it please.

jhechtf commented:
i've checked the actual UX on the preview deploy, which I'm fine with. I just need to get through the code itself. will try to do that by EOD today.

jhechtf commented on apps/web/src/app/[locale]/challenge/_components/sheet-content-custom.tsx:
```diff
@@ -0,0 +1,115 @@
+  const { allChallenges } = useAllChallengesContext();
+
+  const handleValueChange = (value: string) => {
+    if (value === 'popular') {
```
This would likely be a good candidate for a switch statement, or perhaps some updates in the logic, such as a map of `value` attributes to the requisite objects which could clean this up a bit.

jhechtf commented on apps/web/src/app/[locale]/challenge/_components/sheet-content-custom.tsx:
```diff
@@ -0,0 +1,115 @@
+    })
+    .map((challenge) => challenge.id);
+
+  const inProgressChallenges = challenges
```
These are being calculated on every render, consider wrapping them in a `useMemo`.

jhechtf commented on apps/web/src/app/[locale]/tracks/_components/enroll-button.tsx:
```diff
@@ -14,12 +15,18 @@ interface EnrollButtonProps {
       disabled={isLoading}
       onClick={async () => {
         setIsLoading(true);
+        if (text === 'Enroll') {
```
can be changed
```suggestion
        setIsExplorerDisabled(text === 'Enroll');
```

me-imfhd replied:
fixes in latest push

jhechtf commented on apps/web/src/app/layout.tsx:
```diff
@@ -51,7 +52,7 @@ export const metadata = {
 };
 
-export default function RootLayout({ children }: { children: React.ReactNode }) {
+export default async function RootLayout({ children }: { children: React.ReactNode }) {
```
any reason for the change? based on the content below I'm not sure this is needed.

me-imfhd replied:
sorry for the inconvinience, i agree, it shouldn't be there

me-imfhd commented:
@jhechtf tried to fixed those issue, you mentioned, can you review it now.

jhechtf reviewed (approved):
LGTM
